docs/sc-memory/api/cpp/guides/simple_guide_for_implementing_agent.md

```diff
@@ -0,0 +1,532 @@
+
+1. Write input (initial) construction and output (result) construction of your future agent in SC-code.
+2. Create folder with source and header files for sc-agent implementation.
+3. Also you need write `CMakeLists.txt` file. We use cmake to build projects in C++.
+4. In header file, define a class in C++ for this agent and specifies in it at least class of actions that this agent performs and its program. In such class you can also specify primary initiation condition, initiation condition, and result condition.
+5. In source file, implement all declared methods of agent's class. You can also implement your own methods and use them in an agent program. You can use all C++ and OOP tools as much as possible. 
+6. Create file and implement class for keynodes used by implemented agent.
+7. Implement class for module for subscribing implemented agent.
+8. Write tests for implemented agent.
+
+---
+
+Let's consider an example of implementing an agent to count power of the given set:
+
+### **1. Specify inputs and outputs of your future agent.**
+
+The initial construction of agent might look like this:
+
+```scs
+..action
+<- action_calculate_set_power;
+<- action_initiated;
+-> rrel_1: ..some_set;;
+
+..some_set
+-> ..element_1;
+-> ..element_2;
+-> ..element_3;;
+```
+
+<image src="../images/agents/agent_calculate_set_power_input_construction.png"></image>
+
+The result construction of agent might look like this:
+
+```scs
+..action
+<- action_calculate_set_power;
+<- action_initiated;
+<- action_finished_successfully;
+-> rrel_1: ..some_set;
+=> nrel_result: ..result;;
+
+..result = [*
+  ..some_set => nrel_set_power: [3];;
+*];;
+```
+
+<image src="../images/agents/agent_calculate_set_power_output_construction.png"></image>
+
+In addition to agents that initiate actions themselves and then perform these actions, there is a need to implement agents that perform actions initiated by other agents. For this class of agents, it is much easier to create a initial initiation construction in the knowledge base.
+
+--- 
+
+### **2. Create folder with source and header files for agent and `CMakeLists.txt.` file.**
+
+You should get something like this structure:
+
+```plaintext
+set-agents-module/
+├── CMakeLists.txt
+├── agent/
+│   ├── sc_agent_calculate_set_power.hpp
+│   └── sc_agent_calculate_set_power.сpp
+```
+
+---
+
+### **3. Write `CMakeLists.txt` file.**
+  
+The `CMakeLists.txt` file should describe a process of building your agent code. It should contain various instructions and parameters needed to compile and link agent source code to sc-machine libraries.
+
+**CMakeLists.txt**
+
+```cmake
+file(GLOB SOURCES CONFIGURE_DEPENDS
+    "*.cpp" "*.hpp"
+    "agent/*.cpp" "agent/*.hpp"
+)
+
+# Create and link your library with using libraries.
+add_library(set-agents SHARED ${SOURCES})
+target_link_libraries(set-agents LINK_PRIVATE sc-memory)
```

Comment:
```suggestion
target_link_libraries(set-agents LINK_PRIVATE sc-memory  sc-agents-common)
```
when I make this and try to call `utils::IteratorUtils::getAnyByOutRelation(&m_context, {}, {});` I am getting a lot of errors
/usr/bin/ld: ../sc-kpm/sc-agents-common/libsc-agents-common.a(IteratorUtils.cpp.o): warning: relocation against `_ZN10ScKeynodes33nrel_primary_initiation_conditionE' in read-only section `.text'
/usr/bin/ld: ../sc-kpm/sc-agents-common/libsc-agents-common.a(IteratorUtils.cpp.o): relocation R_X86_64_PC32 against symbol `_ZN5utils22ExceptionInvalidParamsD1Ev' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
gmake[2]: *** [set-agents-module/CMakeFiles/set-agents.dir/build.make:116: ../bin/libset-agents.so] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:2175: set-agents-module/CMakeFiles/set-agents.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[100%] Built target sc-server-tests
gmake: *** [Makefile:146: all] Error 2
![image](https://github.com/user-attachments/assets/af3344a3-921d-43cc-9067-66d170d9bc7b)

Reply:
Fixed